clean-modular-code/activity-3/clean-code-activity-3.md

```diff
@@ -0,0 +1,254 @@
+* In [activity 2](../activity-2/clean-code-activity-2), you made your code more DRY ("Don't Repeat Yourself") using documented functions and conditionals. 
+
+In this activity, you will build checks into your workflow to handle data processing "features". 
+
+
+### Real world data processing & workflows and edge cases 
+Real-world data rarely can be imported without "work arounds". You will often find unusual data entries and values you don't expect. Sometimes, these values are documented - for example, a 9999 may represent a missing value in a dataset. Other times, there are typos and other errors in the data that you need to handle. Sometimes, call these unusual values or instances in a dataset or workflow "edge cases".  
+
+Writing robust code that handles unexpected values will make your code run smoothly and fail gracefully. This type of code, which combines functions (or classes) and checks within the functions that handle messy data, will make your code easier to maintain over time. 
+
+:::{tip}
+Using functions, classes, and methods (functions within a class) is a great first step in handling messy data. A function or method provides a modular unit you can test outside of the workflow for the edge cases you may encounter. Also, because a function is a modular unit, you can add elements to handle unexpected processing features as you build your workflow.
+:::
+
+something about debuggers? 
+* https://jupyterlab.readthedocs.io/en/stable/user/debugger.html
+
+## Manage the unexpected 
+
+In this activity, you will apply the following strategies:
+
+* [conditional statements](../checks-conditionals/python-conditionals)
+* try/except blocks
+
+to process the JOSS citation data. 
+
+:::{todo}
+What branch is the lesson with try/except // ask for forgiveness, checks elements in??
+IN THIS PR:
+https://github.com/pyOpenSci/lessons/pull/14/files#diff-7f4ff1b75e85d38f3955cca051e68e8746773c279b34c9a0a400b9c2dc1240ff 
+:::
+
+When you can, try to use the Pythonic approach of asking for forgiveness later (ie use try/except blocks) rather than conditional statements.
+
+```{code-cell} ipython3
+---
+editable: true
+slideshow:
+  slide_type: ''
+---
+# This works but is less pythonic
+def clean_title(title):
+    """Notice that this function checks explicitly to see if it's a list and then processes the data.
+    """
+    if isinstance(title, list):
+        return title[0]
+    return title
+```
+
+## More "pythonic" - ask for forgiveness 
+
+easier to ask for forgiveness
+
+```{code-cell} ipython3
+---
+editable: true
+slideshow:
+  slide_type: ''
+---
+# This is the preferred way to catch an error 
+def clean_title(title):
+    """
+    It's more Pythonic to try first and then ask for forgiveness later. 
+    If you are writing tests this also makes your code easier to test. 
+    """
+    try:
+        return title[0]
+    except (TypeError, IndexError):
```

Comment:
i agree! I do think this content by itself could be a full workshop! 
What if i fleshed out that concept a bit. more in the lesson below (this document is the activity) 

then maybe we could have them spend time looking at some examples of issues in the data and identifying what type of error it may throw. i could make the notebook more interactive. And it could link to the explanations in the other lesson. i think your idea is a great one AND it might be more valuable than forcing them to debug the code honestly. debugging takes time and is super hard too. we have 4 hours and 3 activities.